src/VisualStudio/Core/Def/Implementation/Diagnostics/VisualStudioDiagnosticAnalyzerService.cs

```diff
@@ -348,5 +344,87 @@ private bool IsBuildActive()
                 }
             }
         }
+
+        private sealed class StatusBarUpdater : IDisposable
+        {
+            private readonly IVsStatusbar _statusBar;
+            private readonly IThreadingContext _threadingContext;
+            private readonly uint _totalProjectCount;
+            private readonly string _statusMessageWhileRunning;
+            private readonly string _statusMesageOnCompleted;
+            private readonly Timer _timer;
+
+            private int _analyzedProjectCount;
+            private bool _disposed;
+            private uint _statusBarCookie;
+
+            public StatusBarUpdater(IVsStatusbar statusBar, IThreadingContext threadingContext, string? projectOrSolutionName, uint totalProjectCount)
+            {
+                Contract.ThrowIfFalse(threadingContext.HasMainThread);
+                _statusBar = statusBar;
+                _threadingContext = threadingContext;
+                _totalProjectCount = totalProjectCount;
+
+                _statusMessageWhileRunning = projectOrSolutionName != null
+                    ? string.Format(ServicesVSResources.Running_code_analysis_for_0, projectOrSolutionName)
+                    : ServicesVSResources.Running_code_analysis_for_Solution;
+                _statusMesageOnCompleted = projectOrSolutionName != null
+                    ? string.Format(ServicesVSResources.Code_analysis_completed_for_0, projectOrSolutionName)
+                    : ServicesVSResources.Code_analysis_completed_for_Solution;
+
+                // Set the initial status bar progress and text.
+                _statusBar.Progress(ref _statusBarCookie, fInProgress: 1, _statusMessageWhileRunning, nComplete: 0, nTotal: totalProjectCount);
+                _statusBar.SetText(_statusMessageWhileRunning);
+
+                // Create a timer to periodically update the status message while running analysis.
+                _timer = new Timer(new TimerCallback(UpdateStatusOnTimer), new AutoResetEvent(false),
+                    dueTime: TimeSpan.FromSeconds(5), period: TimeSpan.FromSeconds(5));
+            }
+
+            internal void OnProjectAnalyzed(Project _)
+            {
+                var analyzedProjectCount = Interlocked.Increment(ref _analyzedProjectCount);
+                UpdateStatus(isRunning: analyzedProjectCount < _totalProjectCount);
+            }
+
+            // Add a message to VS status bar that we are running code analysis.
+            private void UpdateStatusOnTimer(object state)
+                => UpdateStatus(isRunning: true);
```

Comment:
i always worry about flurries of events sent to the UI thread.

Reply:
The VS status bar APIs can only be invoked from UI thread - I am open to tuning these status updates to happen only on each project analysis completing and/or just periodically. We do need the updates to happen though for the progress bar to be updated.